feat(langchain): add wrap_model_call (#2673)
langchain-ai/docs: src/oss/langchain/middleware/custom.mdx

### Composition with multiple middleware

When multiple middleware layers return `ExtendedModelResponse`, their commands compose:

- **Inner commands propagate outward:** Outer middleware always receives a plain `ModelResponse` from `handler()`. `ExtendedModelResponse` is unwrapped at composition boundaries.
- **Commands are applied through reducers:** Each `Command` becomes a separate state update. For messages, this means they are additive.
- **Outer wins on conflicts:** For non-reducer state fields, commands are applied inner-first, then outer. The outermost middleware's value takes precedence on conflicting keys.
- **Retry-safe:** When outer middleware retries by calling `handler()` again, accumulated inner commands are cleared and re-collected.

```python
from langchain.agents.middleware import (
    AgentMiddleware,
    ExtendedModelResponse,
    ModelRequest,
    ModelResponse,
)
from langgraph.types import Command
from typing import Callable

class OuterMiddleware(AgentMiddleware):
    def wrap_model_call(
        self,
        request: ModelRequest,
        handler: Callable[[ModelRequest], ModelResponse | ExtendedModelResponse],
    ) -> ExtendedModelResponse:

>         response = handler(request)  # sees ModelResponse, not ExtendedModelResponse

Reviewer: don't need the comment here since signature hasn't changed

:::

## State updates

Both node-style and wrap-style hooks can update agent state. The mechanism differs:

> - **Node-style hooks** (`before_agent`, `before_model`, `after_model`, `after_agent`): Return a dict directly. The dict is merged into the agent state.

Author: Code comment said "State updates to merge into the agent state" - sorry
```suggestion
- **Node-style hooks** (`before_agent`, `before_model`, `after_model`, `after_agent`): Return a dict directly. The dict is applied to the agent state using the graph's reducers.
```

    def wrap_model_call(
        self,
        request: ModelRequest,
        handler: Callable[[ModelRequest], ModelResponse],
    ) -> ExtendedModelResponse:
        response = handler(request)
        return ExtendedModelResponse(
            model_response=response,
            command=Command(update={
                "trace_layer": "outer",
                "messages": [SystemMessage(content="[Outer ran]")],
            }),
        )

class InnerMiddleware(AgentMiddleware):
    """Adds trace_layer and message. Outer adds to same keys; trace_layer: outer wins, messages: additive."""

    def wrap_model_call(
        self,
        request: ModelRequest,
        handler: Callable[[ModelRequest], ModelResponse],
    ):
        response = handler(request)
        return ExtendedModelResponse(
            model_response=response,
            command=Command(update={
                "trace_layer": "inner",
                "messages": [SystemMessage(content="[Inner ran]")],
            }),
        )
```

:::

:::js

```typescript
import * as z from "zod";
import { createMiddleware } from "langchain";
import { Command } from "@langchain/langgraph";

> import { withLangGraph } from "@langchain/langgraph/zod";

Reviewer 2: ```suggestion
import { Command, StateSchema, ReducedValue } from "@langchain/langgraph";
```
